matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/token/HomeserverAccessTokenProvider.kt

```diff
@@ -16,13 +16,107 @@
 
 package org.matrix.android.sdk.internal.network.token
 
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.matrix.android.sdk.api.extensions.orFalse
+import org.matrix.android.sdk.api.failure.Failure
+import org.matrix.android.sdk.api.failure.GlobalError
+import org.matrix.android.sdk.api.failure.MatrixError.Companion.M_FORBIDDEN
+import org.matrix.android.sdk.api.failure.MatrixError.Companion.M_UNKNOWN_TOKEN
+import org.matrix.android.sdk.api.failure.isTokenUnknownError
 import org.matrix.android.sdk.internal.auth.SessionParamsStore
+import org.matrix.android.sdk.internal.auth.data.RefreshResult
+import org.matrix.android.sdk.internal.auth.refresh.RefreshTokenTask
 import org.matrix.android.sdk.internal.di.SessionId
+import org.matrix.android.sdk.internal.network.GlobalErrorReceiver
+import timber.log.Timber
 import javax.inject.Inject
 
 internal class HomeserverAccessTokenProvider @Inject constructor(
         @SessionId private val sessionId: String,
-        private val sessionParamsStore: SessionParamsStore
+        private val sessionParamsStore: SessionParamsStore,
+        private val refreshTokenTask: RefreshTokenTask,
+        private val globalErrorReceiver: GlobalErrorReceiver
 ) : AccessTokenProvider {
```

Comment:
Okhttp has an API to handle accessToken and refreshToken when needed, might be the opportunity to use it instead of a custom interceptor/provider ? 
See this article: https://www.lordcodes.com/articles/authorization-of-web-requests-for-okhttp-and-retrofit for example